homeassistant/components/ws66i/config_flow.py

```diff
@@ -0,0 +1,143 @@
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    ws66i = get_ws66i(data[CONF_IP_ADDRESS])
+    await hass.async_add_executor_job(ws66i.open)
+    # No exception. run a simple test to make sure we opened correct port
+    # Test on FIRST_ZONE because this zone will always be valid
+    ret_val = await hass.async_add_executor_job(ws66i.zone_status, FIRST_ZONE)
+    if ret_val is None:
+        ws66i.close()
+        raise ConnectionError("Not a valid WS66i connection")
+
+    # Validation done. No issues. Close the connection
+    ws66i.close()
+
+    # Return info that you want to store in the config entry.
+    return {CONF_IP_ADDRESS: data[CONF_IP_ADDRESS]}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for WS66i 6-Zone Amplifier."""
+
+    VERSION = 1
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+                return self.async_create_entry(title="WS66i Amp", data=info)
+            except ConnectionError:
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors
+        )
+
+    @staticmethod
+    @core.callback
+    def async_get_options_flow(config_entry):
+        """Define the config flow to handle options."""
+        return Ws66iOptionsFlowHandler(config_entry)
+
+
+@core.callback
+def _key_for_source(index, source, previous_sources):
+    if previous_sources is not None and str(index) in previous_sources:
+        key = vol.Optional(
+            source, description={"suggested_value": previous_sources[str(index)]}
+        )
+    else:
+        key = vol.Optional(source)
+
+    return key
+
+
+class Ws66iOptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle a WS66i options flow."""
+
+    def __init__(self, config_entry):
+        """Initialize."""
+        self.config_entry = config_entry
+
+    async def async_step_init(self, user_input=None):
+        """Manage the options."""
+        if user_input is not None:
+            return self.async_create_entry(
+                title="Source Names",
+                data={CONF_SOURCES: _sources_from_config(user_input)},
+            )
+
+        # Fill form with previous source names
+        previous_sources = self.config_entry.options[CONF_SOURCES]
+        options = {
+            _key_for_source(idx + 1, source, previous_sources): str
+            for idx, source in enumerate(SOURCES)
+        }
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(options),
+        )
```

Comment:
Its a bit unusual to not be able to get the names of the sources from the device

Reply:
The amplifier doesn't provide names of sources. So I set them to "Source 1", "Source 2", etc. And I prefill the options flow with those names, unless user changes them, then I prefill with those. Also, if I don't set default names, sources won't be able to be selected when controlling the amp via HA until they are set.